tools/ami-cleanup/src/internal/cleanup.go

```diff
@@ -0,0 +1,239 @@
+	return getAllWithPagination(
+		func(previousToken *string) (*string, []accountTypes.Region, error) {
+			results, err := client.ListRegions(ctx, &account.ListRegionsInput{
+				RegionOptStatusContains: []accountTypes.RegionOptStatus{
+					accountTypes.RegionOptStatusEnabled,
+					accountTypes.RegionOptStatusEnabledByDefault,
+				},
+				NextToken: previousToken,
+			})
+			if err != nil {
+				return nil, nil, trace.Wrap(err, "failed to request enabled regions")
+			}
+
+			return results.NextToken, results.Regions, nil
+		},
+	)
+}
+
+func (ai *ApplicationInstance) getDevImagesInRegion(ctx context.Context, client IEc2Api) ([]ec2Types.Image, error) {
+	// These are a weird language workaround for getting a pointer to a boolean literal
+	includeTrue := true
+	nameFilterName := "name"
+	stateFilterName := "state"
+	requestInput := &ec2.DescribeImagesInput{
+		Filters: []ec2Types.Filter{
+			{
+				Name:   &nameFilterName,
+				Values: []string{"*dev*"},
+			},
+			{
+				Name:   &stateFilterName,
+				Values: []string{string(ec2Types.ImageStateAvailable)},
+			},
+		},
+		IncludeDeprecated: &includeTrue,
+		IncludeDisabled:   &includeTrue,
+		Owners:            []string{"self"},
+	}
+
+	return getAllWithPagination(
+		func(previousToken *string) (*string, []ec2Types.Image, error) {
+
+			requestInput.NextToken = previousToken
+			results, err := client.DescribeImages(ctx, requestInput)
+			if err != nil {
+				return nil, nil, trace.Wrap(err, "failed to request a dev images")
+			}
+
+			return results.NextToken, results.Images, nil
+		},
+	)
+}
+
+// Repeatedly calls `action` until the returned `nextToken` is null, and accumulates the results.
+// Useful for AWS API calls who's results may be paginated.
+func getAllWithPagination[T any](action func(previousToken *string) (nextToken *string, results []T, err error)) ([]T, error) {
+	var previousToken *string
+	var results []T
+	for {
+		nextToken, newResults, err := action(previousToken)
+		if err != nil {
+			return results, trace.Wrap(err, "failed to get the next set of results")
+		}
+
+		results = append(results, newResults...)
+		if nextToken == nil {
+			return results, nil
+		}
+		previousToken = nextToken
+	}
+}
+
+func (ai *ApplicationInstance) cleanupImageIfOld(ctx context.Context, client IEc2Api, image ec2Types.Image) (int32, error) {
+	creationDate, err := iso8601.ParseString(*image.CreationDate)
+	if err != nil {
+		return 0, trace.Wrap(err, "failed to parse image %q creation timestamp %q as an ISO 8601 value", *image.Name, *image.CreationDate)
+	}
+
+	// If the image is less than a month old, don't do anything
+	imageAge := time.Since(creationDate)
+	if imageAge <= 30*24*time.Hour {
+		return 0, nil
+	}
+
+	log.Printf("\tDeleting %s old AMI %q", imageAge.String(), *image.Name)
+	_, err = client.DeregisterImage(ctx, &ec2.DeregisterImageInput{
+		ImageId: image.ImageId,
+		DryRun:  &ai.shouldDoDryRun,
+	})
+	if err != nil && (!IsDryRunError(err) || !ai.shouldDoDryRun) {
```

Comment:
Why are we checking for `!ai.shouldDoDryRun` here?

Reply:
The AWS API will return a "dry run error" upon success, which needs special handling to ignore.